kappa.js

```diff
@@ -0,0 +1,319 @@
+
+  use (name, source, view, opts = {}) {
+    opts.status = opts.status || this.status
+    opts.context = opts.context || this
+    const flow = new Flow(name, source, view, opts)
+    this.flows[name] = flow
+    this.view[name] = flow.view
+    this.source[name] = flow.source
+    flow.on('error', err => this.emit('error', err, flow))
+
+    this.emit('flow', name)
+
+    if (this.status !== Status.Paused) flow.open()
+
+    return flow
+  }
+
+  pause () {
+    this.status = Status.Paused
+    Object.values(this.flows).forEach(flow => flow.pause())
+  }
+
+  resume () {
+    if (this.status !== Status.Paused) return
+    Object.values(this.flows).forEach(flow => flow.resume())
+    this.status = Status.Ready
+  }
+
+  reset (name, cb) {
+    const flow = this.flows[name]
+    if (!flow) return cb(new Error('Unknown flow: ' + name))
+    flow.reset(cb)
+  }
+
+  ready (names, cb) {
+    if (typeof names === 'function') return this.ready(null, names)
+    if (typeof names === 'string') names = [names]
+    if (!names) names = Object.keys(this.flows)
+
+    let pending = names.length
+    for (const name of names) {
+      const flow = this.flows[name]
+      if (!flow) return cb(new Error('Unknown flow: ' + name))
+      flow.ready(done)
+    }
+    function done () {
+      if (--pending === 0) cb()
+    }
+  }
+
+  close (cb) {
+    let flows = Object.values(this.flows)
+    let pending = flows.length
+    if (!pending) done()
+    // TODO: Propagate errors?
+    flows.forEach(flow => flow.close(done))
+    function done () {
+      if (--pending === 0) cb()
+    }
+  }
+}
+
+class Flow extends EventEmitter {
+  constructor (name, source, view, opts) {
+    super()
+
+    this.opts = opts
+    this.name = name
+
+    this._view = view
+    this._source = source
+
+    this.context = opts.context
+    this.status = opts.status || Status.Ready
+
+    // Assign view and source apis
+    this.view = {}
+    this.source = {}
+    this.view.ready = cb => this.ready(cb)
+    if (view.api) {
+      for (let [key, value] of Object.entries(view.api)) {
+        this.view[key] = bindFn(value, this, this.context)
+      }
+      delete view.api
+    }
+    if (source.api) {
+      for (let [key, value] of Object.entries(source.api)) {
+        this.source[key] = bindFn(value, this, this.context)
+      }
+      delete source.api
+    }
+
+    // Create the list of funtions through which messages run between pull and map.
+    this._transform = new Pipeline()
+    if (this._source.filter) this._transform.push(this._source.filter.bind(this._source))
+    if (this._source.transform) this._transform.push(this._source.transform.bind(this._source))
+    if (this.opts.filter) this._transform.push(this.opts.filter)
+    if (this.opts.transform) this._transform.push(this.opts.transform)
+    if (this._view.filter) this._transform.push(this._view.filter.bind(this._view))
+    if (this._view.transform) this._transform.push(this._view.transform.bind(this._view))
```

Comment:
The use caes for the transform hook is that at several times, I wanted to have a slightly different version of a source that emits messages in a different format to match what a view expects. With `opts.transform` this can be done for a specific flow, `view.transform` and `map.transform` allows to extend existing views/sources and adapt them. `filter` was just the same so I removed this. I think the little added complexity by the transform step is worth the gained flexibility in composition. I'll rethink this another time soon when I go over my existing uses once more to reflect on the APIs.